### PR TITLE
Unify sentiment documentation

### PR DESCRIPTION
**Pull request by vwxyzjn**

### CODE REVIEW

File: docs/source/example_overview.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Examples
+
+_The best place to learn about examples in TRL is our [docs page](https://huggingface.co/docs/trl/index)!_
```
Comment (lvwerra) [inline docs/source/example_overview.md:3]: Since people are already in the docs this is not necessary, right?

```suggestion
```

File: docs/source/example_overview.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+## Introduction
+
+The examples should work in any of the following settings (with the same script):
+   - single CPU or single GPU
+   - multi GPUS (using PyTorch distributed mode)
+   - multi GPUS (using DeepSpeed ZeRO-Offload stages 1 & 2)
+   - fp16 (mixed-precision) or fp32 (normal precision)
```
Comment (lvwerra) [inline docs/source/example_overview.md:11]: also bf16

File: docs/source/example_overview.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+To run it in each of these various modes, first initialize the accelerate
+configuration with `accelerate config`
+
+**NOTE for to train with a 8-bit model a more recent version of**
+transformers is required, for example:
```
Comment (lvwerra) [inline docs/source/example_overview.md:17]: The formatting of this is a bit odd. Maybe also mention 4-bit. Also the installation command actually installs bnb and peft and not transformers?

File: docs/source/example_overview.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+You can also adjust the parameters of the 🤗 Accelerate config file to suit your needs (e.g. training in mixed precision).
+
+### Distributed training with DeepSpeed
+
+Most of the scripts can be run on multiple GPUs together with DeepSpeed ZeRO-{1,2,3} for efficient sharding of the optimizer states, gradients, and model weights. To do so, run:
+
+```shell
+accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero{1,2,3}.yaml --num_processes {NUM_GPUS} path_to_script.py --all_arguments_of_the_script
```
Comment (lvwerra) [inline docs/source/example_overview.md:75]: maybe add a note that people can adapt the deepspeed configs to their needs?

File: docs/source/sentiment_tuning.mdx [hunk 5]
```diff
@@ -4,35 +4,129 @@ The notebooks and scripts in this examples show how to fine-tune a model with a
...
+| [`examples/notebooks/gpt2-sentiment.ipynb`](https://github.com/huggingface/trl/tree/main/examples/notebooks/gpt2-sentiment.ipynb)              | This notebook demonstrates how to reproduce the GPT2 imdb sentiment tuning example on a jupyter notebook.                |
+| [`examples/notebooks/gpt2-control.ipynb`](https://github.com/huggingface/trl/tree/main/examples/notebooks/gpt2-control.ipynb)   [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/trl/blob/main/examples/sentiment/notebooks/gpt2-sentiment-control.ipynb)                | This notebook demonstrates how to reproduce the GPT2 sentiment control example on a jupyter notebook.    
+
+
+Here are also some easier-to-run colab notebooks that you can use to get started with TRL:
```
Comment (lvwerra) [inline docs/source/sentiment_tuning.mdx:16]: That should go before the table, no?

**Review by lvwerra — COMMENTED** [review]
Looks good overall, just a few minor fixes/nits!

File: docs/source/sentiment_tuning.mdx (hunk 5 above)
Comment (vwxyzjn) [reply docs/source/sentiment_tuning.mdx:16]: Removed. As above it we already have `Here's an overview of the notebooks and scripts`.

File: docs/source/example_overview.md (hunk 4 above)
Comment (vwxyzjn) [reply docs/source/example_overview.md:74]: Above that there was already "You can also adjust the parameters of the 🤗 Accelerate config file to suit your needs (e.g. training in mixed precision)." I added some more docs

File: docs/source/example_overview.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
+# Examples
+
+
+## Introduction
+
+The examples should work in any of the following settings (with the same script):
+   - single CPU or single GPU
+   - multi GPUS (using PyTorch distributed mode)
+   - multi GPUS (using DeepSpeed ZeRO-Offload stages 1 & 2)
```
Comment (lewtun) [inline docs/source/example_overview.md:9]:
```suggestion
   - multi GPUS (using DeepSpeed ZeRO-Offload stages 1, 2, & 3)
```

File: docs/source/example_overview.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
+# Examples
+
+
+## Introduction
+
+The examples should work in any of the following settings (with the same script):
+   - single CPU or single GPU
```
Comment (lewtun) [inline docs/source/example_overview.md:7]: I don't think the examples can actually run on CPU right? E.g. if we activate 8-bit, the PPO example likely won't work IIRC
```suggestion
   - single GPU
```

File: docs/source/example_overview.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+   - multi GPUS (using PyTorch distributed mode)
+   - multi GPUS (using DeepSpeed ZeRO-Offload stages 1 & 2)
+   - fp16 (mixed-precision), fp32 (normal precision), or bf16 (bfloat16 precision)
+
+To run it in each of these various modes, first initialize the accelerate
+configuration with `accelerate config`
+
+**NOTE for to train with a 4-bit or 8-bit model**, please run
```
Comment (lewtun) [inline docs/source/example_overview.md:15]:
```suggestion
**NOTE to train with a 4-bit or 8-bit model**, please run
```

File: docs/source/example_overview.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+```bash
+pip install --upgrade trl[quantization]
+```
+
+
+## Accelerate Config
+For all the examples, you'll need to generate an `Accelerate` config with:
```
Comment (lewtun) [inline docs/source/example_overview.md:23]:
```suggestion
For all the examples, you'll need to generate a 🤗 Accelerate config file with:
```

File: docs/source/example_overview.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+| [`examples/scripts/sft_trainer.py`](https://github.com/huggingface/trl/blob/main/examples/scripts/sft_trainer.py) | This script shows how to use the `SFTTrainer` to fine tune a model or adapters into a target dataset.                     |
+| [`examples/scripts/reward_trainer.py`](https://github.com/huggingface/trl/blob/main/examples/scripts/reward_trainer.py) | This script shows how to use the `RewardTrainer` to train a reward model on your own dataset.                            |
+| [`examples/scripts/sentiment_tuning.py`](https://github.com/huggingface/trl/blob/main/examples/scripts/sentiment_tuning.py) | This script shows how to use the `PPOTrainer` to fine-tune a sentiment analysis model using IMDB dataset                 |
+| [`examples/scripts/multi_adapter_rl.py`](https://github.com/huggingface/trl/blob/main/examples/scripts/multi_adapter_rl.py) | This script shows how to use the `PPOTrainer` to train a single base model with multiple adapters. Requires you to run the example script with the reward model training beforehand. |
```
Comment (lewtun) [inline docs/source/example_overview.md:40]: Should we eventually migrate the DPO script here as well?

File: docs/source/sentiment_tuning.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+# 2. run via `accelerate` (recommended), enabling more features (e.g., multiple GPUs, deepspeed)
+accelerate config # will prompt you to define the training configuration
+accelerate launch examples/scripts/sentiment_tuning.py # launches training
+# 3. get help text and documentation
+python examples/scripts/sentiment_tuning.py --help
+# 4. configure logging with wandb and, say, mini_batch_size=1 and gradient_accumulation_steps=16
+python examples/scripts/sentiment_tuning.py --ppo_config.log_with wandb --ppo_config.mini_batch_size 1 --ppo_config.gradient_accumulation_steps 16
+```
+
+Note: if you don't want to log with `wandb` remove `log_with="wandb"` in the scripts/notebooks. You can also replace it with your favourite experiment tracker that's [supported by `accelerate`](https://huggingface.co/docs/accelerate/usage_guides/tracking).
```
Comment (lewtun) [inline docs/source/sentiment_tuning.md:31]: Not for this PR, but I think we should set the examples such that `wandb` is opt-in by default

File: docs/source/sentiment_tuning.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+
+## Few notes on multi-GPU 
+
+To run in multi-GPU setup with DDP (distributed Data Parallel) change the `device_map` value to `device_map={"": Accelerator().process_index}` and make sure to run your script with `accelerate launch yourscript.py`. If you want to apply naive pipeline parallelism you can use `device_map="auto"`.
```
Comment (lewtun) [inline docs/source/sentiment_tuning.md:36]: This only applies for PEFT trainings AFAIK - for DeepSpeed you don't want to specify the device map explicitly. W

File: docs/source/sentiment_tuning.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+
+## Benchmark
```
Comment (lewtun) [inline docs/source/sentiment_tuning.md:39]:
```suggestion
## Benchmarks
```

File: docs/source/sentiment_tuning.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    --slurm-ntasks 1 \
+    --slurm-total-cpus 12 \
+    --slurm-template-path benchmark/trl.slurm_template
+```
+
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/benchmark/v0.4.7-55-g110e672/sentiment.png)
+
+
+
+## w/ and w/o gradient accumulation
```
Comment (lewtun) [inline docs/source/sentiment_tuning.md:60]:
```suggestion
## With and without gradient accumulation
```

File: docs/source/sentiment_tuning.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    --slurm-gpus-per-task 1 \
+    --slurm-ntasks 1 \
+    --slurm-total-cpus 12 \
+    --slurm-template-path benchmark/trl.slurm_template
+```
+
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/benchmark/v0.4.7-55-g110e672/gradient_accu.png)
+
+
+## w/ different models (gpt2, gpt2-xl, falcon, llama2)
```
Comment (lewtun) [inline docs/source/sentiment_tuning.md:77]:
```suggestion
## Comparing different models (gpt2, gpt2-xl, falcon, llama2)
```

File: docs/source/sentiment_tuning.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    --slurm-nodes 1 \
+    --slurm-gpus-per-task 1 \
+    --slurm-ntasks 1 \
+    --slurm-total-cpus 12 \
+    --slurm-template-path benchmark/trl.slurm_template
+```
+
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/benchmark/v0.4.7-55-g110e672/different_models.png)
+
+## w/ and w/o PEFT
```
Comment (lewtun) [inline docs/source/sentiment_tuning.md:114]:
```suggestion
## With and without PEFT
```

File: docs/source/sentiment_tuning.mdx [hunk 17]
```diff
@@ -4,35 +4,129 @@ The notebooks and scripts in this examples show how to fine-tune a model with a
...
+python examples/scripts/sentiment_tuning.py --ppo_config.log_with wandb --ppo_config.mini_batch_size 1 --ppo_config.gradient_accumulation_steps 16
 ```
 
 Note: if you don't want to log with `wandb` remove `log_with="wandb"` in the scripts/notebooks. You can also replace it with your favourite experiment tracker that's [supported by `accelerate`](https://huggingface.co/docs/accelerate/usage_guides/tracking).
 
 
-## Launch scripts
+## Few notes on multi-GPU 
+
+To run in multi-GPU setup with DDP (distributed Data Parallel) change the `device_map` value to `device_map={"": Accelerator().process_index}` and make sure to run your script with `accelerate launch yourscript.py`. If you want to apply naive pipeline parallelism you can use `device_map="auto"`.
```
Comment (lewtun) [inline docs/source/sentiment_tuning.mdx:38]: Same comment here as above about DeepSpeed vs PEFT

File: docs/source/sentiment_tuning.mdx [hunk 18]
```diff
@@ -4,35 +4,129 @@ The notebooks and scripts in this examples show how to fine-tune a model with a
...
+    --slurm-total-cpus 12 \
+    --slurm-template-path benchmark/trl.slurm_template
 ```
 
-## Few notes on multi-GPU 
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/benchmark/v0.4.7-55-g110e672/sentiment.png)
+
+
+
+## w/ and w/o gradient accumulation
```
Comment (lewtun) [inline docs/source/sentiment_tuning.mdx:62]:
```suggestion
## With and without gradient accumulation
```

File: docs/source/sentiment_tuning.mdx [hunk 19]
```diff
@@ -4,35 +4,129 @@ The notebooks and scripts in this examples show how to fine-tune a model with a
...
+    --slurm-gpus-per-task 1 \
+    --slurm-ntasks 1 \
+    --slurm-total-cpus 12 \
+    --slurm-template-path benchmark/trl.slurm_template
+```
+
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/benchmark/v0.4.7-55-g110e672/gradient_accu.png)
+
+
+## w/ different models (gpt2, gpt2-xl, falcon, llama2)
```
Comment (lewtun) [inline docs/source/sentiment_tuning.mdx:80]:
```suggestion
## Comparing different models (gpt2, gpt2-xl, falcon, llama2)
```

File: docs/source/sentiment_tuning.mdx [hunk 20]
```diff
@@ -4,35 +4,129 @@ The notebooks and scripts in this examples show how to fine-tune a model with a
...
+    --slurm-nodes 1 \
+    --slurm-gpus-per-task 1 \
+    --slurm-ntasks 1 \
+    --slurm-total-cpus 12 \
+    --slurm-template-path benchmark/trl.slurm_template
+```
+
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/benchmark/v0.4.7-55-g110e672/different_models.png)
+
+## w/ and w/o PEFT
```
Comment (lewtun) [inline docs/source/sentiment_tuning.mdx:117]:
```suggestion
## With and without PEFT
```

File: docs/source/sentiment_tuning.md
Comment (lewtun) [inline docs/source/sentiment_tuning.md:1]: Why is there a `.md` and `.mdx` file? I think there should only be one right?

**Review by lewtun — APPROVED** [review]
Thanks for adding this @vwxyzjn ! Looks great and I left only a few minor nits

File: docs/source/example_overview.md (hunk 10 above)
Comment (vwxyzjn) [reply docs/source/example_overview.md:40]: Yes, DPO / any other examples should be documented here eventually. Probably similar to https://docs.cleanrl.dev/rl-algorithms/overview/

File: docs/source/sentiment_tuning.md (hunk 12 above)
Comment (vwxyzjn) [reply docs/source/sentiment_tuning.md:36]: Good point. Imo we should just have examples like in the benchmark scripts and handle these situations ourselves.

File: docs/source/sentiment_tuning.md
Comment (vwxyzjn) [reply docs/source/sentiment_tuning.md:1]: I have just removed the `.md`. I initially renamed `.mdx` to `.md` so syntax highlighting worked again but forgot to remove it later 🙈. 

image

**Review by lvwerra — APPROVED** [review]
Just the last comment about adding the dependencies to `[dev]` recipe, then we can merge in my opinion :)

File: setup.py [hunk 21]
```diff
@@ -71,9 +71,16 @@
...
+    "dev": [
+        "parameterized",
+        "pytest",
+        "pytest-xdist",
+        "pre-commit",
+        "peft>=0.4.0",
+        "diffusers>=0.18.0",
+        "deepspeed>=0.9.5",
+    ],
+    "benchmark": ["wandb", "ghapi", "openrlbenchmark==0.2.1a5", "requests", "deepspeed", "bitsandbytes>=0.41.0"],
```
Comment (lvwerra) [inline setup.py:83]: i think a small misunderstanding :) what i meant with my previous comment is that we can have a quantization and a deepspeed recipe but the dev recipe is ideally always the union of all recipes. does that make sense?

Comment (vwxyzjn) [reply setup.py]: Ah my bad. Just updated.
